### PR TITLE
session: improvment refactor the logic of load data batch insert

### PR DESCRIPTION
**Pull request by cfzjywxk**
…ce per transaction

### What problem does this PR solve? 

issue 3092 
load data batch checked all rows for one packet instead of one transaction, will cause too many batch-get to the storage node.
We could do one batch-get only for one transaction.

local rough tests show that
batch get per transaction may not be always better,  
the cost depends on (batchget efficiency with more keys ) and batchget call rt
generally, slower batchget calls result in more load data time cost

### What is changed and how it works?

make batchget happens one per transcation but not per packet

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

### CODE REVIEW

File: executor/load_data.go [hunk 1]
```diff
@@ -261,14 +262,29 @@ func (e *LoadDataInfo) InsertData(prevData, curData []byte) ([]byte, bool, error
...
+		if err != nil {
+			return nil, reachLimit, err
+		}
 	}
 	return curData, reachLimit, nil
 }
 
+// CheckAndInsertOneBatch is used to commit one transaction batch full filled data
+func (e *LoadDataInfo) CheckAndInsertOneBatch() error {
+	var err error
```
Comment (crazycs520) [inline executor/load_data.go]: how about 
```go
	if len(e.rows) == 0 {
		return nil
	}
	e.ctx.GetSessionVars().StmtCtx.AddRecordRows(uint64(len(e.rows)))
	err := e.batchCheckAndInsert(e.rows, e.addRecordLD)
	e.rows = e.rows[:0]
	return err
```

**Comment by winoros** [comment]
The title is a little long. Please shorten it.

File: executor/load_data.go [hunk 2]
```diff
@@ -213,7 +215,7 @@ func (e *LoadDataInfo) getLine(prevData, curData []byte) ([]byte, []byte, bool)
 // If it has the rest of data isn't completed the processing, then it returns without completed data.
 // If the number of inserted rows reaches the batchRows, then the second return value is true.
 // If prevData isn't nil and curData is nil, there are no other data to deal with and the isEOF is true.
-func (e *LoadDataInfo) InsertData(prevData, curData []byte) ([]byte, bool, error) {
+func (e *LoadDataInfo) InsertData(prevData, curData []byte, txnBatchCheckIns bool) ([]byte, bool, error) {
```
Comment (lysu) [inline executor/load_data.go]: maybe better add some comment for `txnBatchCheckIns` argument

File: executor/load_data.go [hunk 3]
```diff
@@ -261,14 +262,29 @@ func (e *LoadDataInfo) InsertData(prevData, curData []byte) ([]byte, bool, error
 			break
 		}
 	}
-	e.ctx.GetSessionVars().StmtCtx.AddRecordRows(uint64(len(rows)))
-	err := e.batchCheckAndInsert(rows, e.addRecordLD)
-	if err != nil {
-		return nil, reachLimit, err
+	if !txnBatchCheckIns {
+		err := e.CheckAndInsertOneBatch()
```
Comment (lysu) [inline executor/load_data.go]: after this PR, it seems only unit-test logic will step into here, how about this code snippet  and remove `txnBatchCheckIns` let it always be `true` and modify unit-test to test real check logic

Comment (cfzjywxk) [reply executor/load_data.go]: @lysu  @jackysp 
some local rought tests(see jira issue 3092 comments) show that batchget per packet may not always be fasttter,  should we make it configurable to use batch per transaction or packet?

**Comment by cfzjywxk** [comment]
@jackysp @lysu PTAL

**Comment by lysu** [comment]
@cfzjywxk this pr need cherry-pick to 3.0

**Pull request merged**